### PR TITLE
Add <[T; N]>::copied

### PR DESCRIPTION
**Pull request by scottmcm**
This works (it's a nicer version of how I did things in `ArrayTools`), so I figured I'd just send a PR and see what people think about it.

r? @LukasKalbertodt 
cc https://github.com/rust-lang/rust/issues/76118

### CODE REVIEW

**Comment by LukasKalbertodt** [comment]
I *really* want that method (and `cloned`), but I would prefer not using this kind of workaround :/
I think the correct solution would be like `Option<T>` does it: `impl<T: Copy> Option<&T> { ... }`. For that we need a way to add multiple impl blocks, which is not currently possible with only one lang item. I started [a discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Multiple.20impl.20blocks.20for.20built-in.20types.20(lang.20items.3F)) about this. Hopefully T-lang can help out :)

File: library/core/src/array/mod.rs [hunk 1]
```diff
@@ -415,11 +415,98 @@ impl<T, const N: usize> [T; N] {
...
+    pub fn copied(&self) -> [<T as Reference>::Pointee; N]
+    where
+        T: Reference<Pointee: Copy>,
+    {
+        // No guard needed here because these are all bitwise copies, which cannot fail.
+        // (And things that are `Copy` can't be `Drop` anyway.)
+        let mut dst = MaybeUninit::uninit_array::<N>();
+        for i in 0..N {
+            dst[i].write(self[i].copy());
+        }
```
Comment (matklad) [inline library/core/src/array/mod.rs:459]: Out of curiosity, do we get "guaranteed" bounds check elimination here? That is, am I correct that rewriting this as `dst.iter_mut().zip(self.iter())` would make this strictly worse?

Comment (scottmcm) [reply library/core/src/array/mod.rs:459]: [After mir-opt](https://github.com/rust-lang/rust/blob/80fc9b0ecb29050d45b17c64af004200afd3cfc2/compiler/rustc_mir/src/transform/instcombine.rs#L64-L67) the check is against a constant, which is quite easy for LLVM to remove.  (The `for i in 0..N` loop is easily canonicalized, especially now that `Range<usize>::next` uses nowrap add.)

Comment (pickfire) [reply library/core/src/array/mod.rs:459]: Must we write byte by byte if it implements `copy`?

**Comment by pickfire** [comment]
I believe the next thing we will be adding is `cloned`?

**Comment by scottmcm** [comment]
Given https://github.com/rust-lang/rust/pull/76236#issuecomment-685419589 I'll just close this.

**Pull request closed without merging**